### PR TITLE
Migrate to flit_core build system

### LINKED ISSUE #357: Please consider using flit build system to ease setuptools bootstrap
I'm the @Gentoo maintainer for core Python packages. We're currently considering options for bootstrapping setuptools while unvendoring its dependencies in the PEP 517 world. One of the interesting options is to use [flit](https://flit.readthedocs.io/en/latest/) as it has much fewer dependencies than setuptools and we've got its bootstrap trivially covered.

Using flit as pyparsing's build system would make bootstrapping setuptools much easier for us. Flit can also generate a backwards-compatible `setup.py` (or be used alongside `setup.py`, though for this to work for us `pyproject.toml` would have to specify flit_core as the `build-backend`). FWICS it is becoming increasingly popular these days.

Would you be interested in this? If yes, then I can prepare a PR either converting the project to flit entirely, or adding it alongside existing `setup.py`, as you prefer.

### PR DESCRIPTION
**Pull request by mgorny**
Fixes #357

### CODE REVIEW

**Comment by mgorny** [comment]
sdist is mostly the same, except for removing setuptools-specific files, adding `pyproject.toml` and changes in metadata.

```
Only in old: MANIFEST.in
diff -dupr old/PKG-INFO pyparsing-3.0.7/PKG-INFO
--- old/PKG-INFO	2022-01-31 10:53:11.057560700 +0100
+++ pyparsing-3.0.7/PKG-INFO	1970-01-01 01:00:00.000000000 +0100
@@ -1,13 +1,10 @@
 Metadata-Version: 2.1
 Name: pyparsing
 Version: 3.0.7
-Summary: Python parsing module
-Home-page: https://github.com/pyparsing/pyparsing/
-Download-URL: https://pypi.org/project/pyparsing/
-Author: Paul McGuire
-Author-email: ptmcg.gm+pyparsing@gmail.com
-License: MIT License
-Platform: UNKNOWN
+Summary: pyparsing module - Classes and methods to define and execute parsing grammars
+Author-email: Erik Rose <erikrose@grinchcentral.com>
+Requires-Python: >=3.6
+Description-Content-Type: text/x-rst
 Classifier: Development Status :: 5 - Production/Stable
 Classifier: Intended Audience :: Developers
 Classifier: Intended Audience :: Information Technology
@@ -23,10 +20,10 @@ Classifier: Programming Language :: Pyth
 Classifier: Programming Language :: Python :: 3 :: Only
 Classifier: Programming Language :: Python :: Implementation :: CPython
 Classifier: Programming Language :: Python :: Implementation :: PyPy
-Requires-Python: >=3.6
-Description-Content-Type: text/x-rst
+Requires-Dist: railroad-diagrams ; extra == "diagrams"
+Requires-Dist: jinja2 ; extra == "diagrams"
+Project-URL: Homepage, https://github.com/pyparsing/pyparsing/
 Provides-Extra: diagrams
-License-File: LICENSE
 
 PyParsing -- A Python Parsing Module
 ====================================
@@ -105,4 +102,3 @@ See `CHANGES <https://github.com/pyparsi
 .. |Coverage| image:: https://codecov.io/gh/pyparsing/pyparsing/branch/master/graph/badge.svg
   :target: https://codecov.io/gh/pyparsing/pyparsing
 
-
Only in old: pyparsing.egg-info
Only in pyparsing-3.0.7: pyproject.toml
Only in old: setup.cfg
Only in old: setup.py
```

**Comment by mgorny** [comment]
wheel only differs in metadata:

```
diff -dupr a/pyparsing-3.0.7.dist-info/METADATA b/pyparsing-3.0.7.dist-info/METADATA
--- a/pyparsing-3.0.7.dist-info/METADATA	2022-01-31 09:53:18.000000000 +0100
+++ b/pyparsing-3.0.7.dist-info/METADATA	2016-01-01 00:00:00.000000000 +0100
@@ -1,13 +1,10 @@
 Metadata-Version: 2.1
 Name: pyparsing
 Version: 3.0.7
-Summary: Python parsing module
-Home-page: https://github.com/pyparsing/pyparsing/
-Download-URL: https://pypi.org/project/pyparsing/
-Author: Paul McGuire
-Author-email: ptmcg.gm+pyparsing@gmail.com
-License: MIT License
-Platform: UNKNOWN
+Summary: pyparsing module - Classes and methods to define and execute parsing grammars
+Author-email: Erik Rose <erikrose@grinchcentral.com>
+Requires-Python: >=3.6
+Description-Content-Type: text/x-rst
 Classifier: Development Status :: 5 - Production/Stable
 Classifier: Intended Audience :: Developers
 Classifier: Intended Audience :: Information Technology
@@ -23,12 +20,10 @@ Classifier: Programming Language :: Pyth
 Classifier: Programming Language :: Python :: 3 :: Only
 Classifier: Programming Language :: Python :: Implementation :: CPython
 Classifier: Programming Language :: Python :: Implementation :: PyPy
-Requires-Python: >=3.6
-Description-Content-Type: text/x-rst
-License-File: LICENSE
+Requires-Dist: railroad-diagrams ; extra == "diagrams"
+Requires-Dist: jinja2 ; extra == "diagrams"
+Project-URL: Homepage, https://github.com/pyparsing/pyparsing/
 Provides-Extra: diagrams
-Requires-Dist: railroad-diagrams ; extra == 'diagrams'
-Requires-Dist: jinja2 ; extra == 'diagrams'
 
 PyParsing -- A Python Parsing Module
 ====================================
@@ -107,4 +102,3 @@ See `CHANGES <https://github.com/pyparsi
 .. |Coverage| image:: https://codecov.io/gh/pyparsing/pyparsing/branch/master/graph/badge.svg
   :target: https://codecov.io/gh/pyparsing/pyparsing
 
-
diff -dupr a/pyparsing-3.0.7.dist-info/RECORD b/pyparsing-3.0.7.dist-info/RECORD
--- a/pyparsing-3.0.7.dist-info/RECORD	2022-01-31 09:53:18.000000000 +0100
+++ b/pyparsing-3.0.7.dist-info/RECORD	2016-01-01 00:00:00.000000000 +0100
@@ -11,7 +11,6 @@ pyparsing/util.py,sha256=kq772O5YSeXOSdP
 pyparsing/diagram/__init__.py,sha256=yySG7RAh6JHuM8xewjaZjY4EWlIc6bX6neHxzTOjuoM,22136
 pyparsing/diagram/template.jinja2,sha256=SfQ8SLktSBqI5W1DGcUVH1vdflRD6x2sQBApxrcNg7s,589
 pyparsing-3.0.7.dist-info/LICENSE,sha256=ENUSChaAWAT_2otojCIL-06POXQbVzIGBNRVowngGXI,1023
-pyparsing-3.0.7.dist-info/METADATA,sha256=u3H31Y4UAq3IJaMyDOjbs1vOOtXaEaSlX2yjCt6NXp8,4227
-pyparsing-3.0.7.dist-info/WHEEL,sha256=G16H4A3IeoQmnOrYV4ueZGKSjhipXx8zc8nu9FGlvMA,92
-pyparsing-3.0.7.dist-info/top_level.txt,sha256=eUOjGzJVhlQ3WS2rFAy2mN3LX_7FKTM5GSJ04jfnLmU,10
+pyparsing-3.0.7.dist-info/WHEEL,sha256=jPMR_Dzkc4X4icQtmz81lnNY_kAsfog7ry7qoRvYLXw,81
+pyparsing-3.0.7.dist-info/METADATA,sha256=1F3bj5RyWPn9K7k7XGWljwu2eknhsrlPe8_F4SPHSjs,4172
 pyparsing-3.0.7.dist-info/RECORD,,
Only in a/pyparsing-3.0.7.dist-info: top_level.txt
diff -dupr a/pyparsing-3.0.7.dist-info/WHEEL b/pyparsing-3.0.7.dist-info/WHEEL
--- a/pyparsing-3.0.7.dist-info/WHEEL	2022-01-31 09:53:18.000000000 +0100
+++ b/pyparsing-3.0.7.dist-info/WHEEL	2016-01-01 00:00:00.000000000 +0100
@@ -1,5 +1,4 @@
 Wheel-Version: 1.0
-Generator: bdist_wheel (0.37.1)
+Generator: flit 3.6.0
 Root-Is-Purelib: true
 Tag: py3-none-any
-
```

**Comment by mgorny** [comment]
I've tried to preserve the original metadata and sdist contents. One notable difference is description — flit grabs that from `pyparsing/__init__.py` docstring by default. If you prefer, I can put the old description in explicitly but I've figured out they roughly mean the same.

I haven't touched `makeRelease.bat` yet as I don't know what environment it's run on and what workflow you prefer. Long story short, two typical options are installing [flit](https://pypi.org/project/flit/) and using `flit build` to create the distribution, or installing [build](https://pypi.org/project/build/) and using `python -m build`. I've been using the latter.

**Comment by ptmcg** [comment]
makeRelease.bat can be deleted, I used it when I still building pyparsing on my Windows machine, which I have not done for years. It is way out of date. I do the tox and twine commands manually now, that's really all the automation I need.

In its place, though, could you add a BUILDING file with instructions on testing and building for release?

**Comment by mgorny** [comment]
Added. Is what you had in mind?

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,53 @@
+[build-system]
+requires = ["flit_core >=3.2,<4"]
+build-backend = "flit_core.buildapi"
+
+[project]
+name = "pyparsing"
+authors = [{name = "Erik Rose", email = "erikrose@grinchcentral.com"}]
```
Comment (ptmcg) [inline pyproject.toml:7]: Copy/pasta fail, I assume - please fix.

**Comment by ptmcg** [comment]
BUILDING file looks great

File: pyproject.toml (hunk 1 above)
Comment (mgorny) [reply pyproject.toml:7]: Such a shameful fail, I'm sorry about that.

**Comment by mgorny** [comment]
Updated the initial commit and force-pushed.

**Comment by ptmcg** [comment]
Sorry, I didn't notice that merging the other PR would generate a conflict here. If you can resolve, I'll go ahead and merge next.

**Comment by mgorny** [comment]
Rebased.

**Comment by mgorny** [comment]
Thanks!

**Pull request merged**